libobs: Improve monitoring deduplication

pkviet opened the pull request:
### Description
The monitoring deduplication was previously checking at each audio tick
the whole audio tree for Audio Output Capture (AOC) devices used for
monitoring. The profiling did not show any big impact on the audio
callback. But due to  reports of a significant slowdown for some users
 we have implemented a new source call triggered whenever:
- a monitoring device is changed in Settings > Audio,
- an Audio Output Capture source changes its device or on startup or is removed;

The main call compares the AOC device with the monitoring device.
If they are identical, a signal is sent to add in audio thread an
audio task updating the AOC duplicating source pointer.
This triggers monitoring deduplication if the ptr is not NULL.

The rest of the logic in obs-audio.c is the same as before these tweaks.

Note that we discovered & fixed also a bug with the function retrieving the default
monitoring device on macOS. It was picking the default device used for
system sounds instead of the default play device.

### Motivation and Context
Improve the previous bugfix.
Fix also for the default monitoring device on macOS.
The bug was introduced in: https://github.com/obsproject/obs-studio/commit/e006d961a479699921323dc1350939221dc0ed58
Just an 8 yrs old bug. Wink to @admshao 

### How Has This Been Tested?
Tested on windows 11, linux ubuntu 24.04 & macos 26.1.
Monitoring deduplication works & the audio callback is faster.
Checked by recording a sine wave which is monitored and duplicated on Desktop Audio; the wav file has the expected level after checking it with ffmpeg volumedetect filter.

### Types of changes
- Bug fix (non-breaking change which fixes an issue)
- Tweak (non-breaking change to improve existing functionality) 
- Performance enhancement (non-breaking change which improves efficiency)

### Checklist:
- [x] My code has been run through [clang-format](https://github.com/obsproject/obs-studio/blob/master/.clang-format).
- [x] I have read the [**contributing** document](https://github.com/obsproject/obs-studio/blob/master/CONTRIBUTING.rst).
- [x] My code is not on the master branch.
- [x] The code has been tested.
- [x] All commit messages are properly formatted and commits squashed where appropriate.
- [x] I have included updates to all appropriate documentation.

prgmitchell commented:
Scene collection with 6 audio sources and nothing else showing immediate and growing rendering lag on Windows 11 25H2 running OBS 32.0.3. Just reporting here to say that this build does indeed fix that.

PatTheMav commented on plugins/mac-capture/mac-audio.c:
```diff
@@ -784,6 +784,9 @@ static void coreaudio_destroy(void *data)
 
 	if (ca) {
 		coreaudio_shutdown(ca);
+		// If the device is also used for monitoring, a cleanup is needed.
```
```suggestion
		/* If the device is also used for monitoring, a cleanup is needed. */
```

pkviet replied:
done

pkviet commented on plugins/linux-pulseaudio/pulse-input.c:
```diff
@@ -523,12 +528,15 @@ static void pulse_update(void *vptr, obs_data_t *settings)
-	new_device = obs_data_get_string(settings, "device_id");
+	const char *new_device = obs_data_get_string(settings, "device_id");
 	if (!data->device || strcmp(data->device, new_device) != 0) {
+		// Signal to deduplication logic in case the device is also used for monitoring.
```
done

RytoEX reviewed (changes requested):
Nits.

pkviet commented on libobs/obs-source.c:
```diff
@@ -368,6 +368,54 @@ static void obs_source_init_audio_hotkeys(struct obs_source *source)
+		calldata_set_ptr(&cd, "source", src);
+		signal_handler_signal(obs->signals, "deduplication_changed", &cd);
+		blog(LOG_INFO,
+		     "Device for 'Audio Output Capture' source %s is also used for audio  monitoring:"
```
i removed some text and left out the colon, my bad. Fixed

RytoEX reviewed (changes requested):
Additional commit message nit: when splitting "paragraphs" in the commit message body, use a blank line between them and after any lists.

pkviet commented:
> Additional commit message nit: when splitting "paragraphs" in the commit message body, use a blank line between them and after any lists.

done

pkviet commented:
> I tested the downloaded artifacts. With this change, using the following configuration (Monitor and Output) causes Scene 2 audio to not be included in the output.
> 
> ```
>  - scene 'Scene':
>     - source: 'Audio Output Capture' (wasapi_output_capture)
>  - scene 'Scene 2':
>     - source: 'Media Source' (ffmpeg_source)
>         - monitoring: monitor and output
> ```
> 
> Both the Audio Output Capture and the monitoring device are set to Default. Note: the output works normally if monitoring is turned off.
> 
> This issue does not occur in version 32.0.3.

I confirm the issue which required a very small fix (a bool test against the active state of the AOC). Test again and it should be ok. Thanks for the report.

pkviet commented:
@Warchamp7 found a regression when switching from 2 scenes with a different AOC. The AOC registered for deduplication is not updated.

pkviet commented:
**Update**
- Fixed the regression. Before the current PR, the monitoring AOC was automatically updated at each audio tick and was active. But with the check of AOC against monitor device being made in UI thread, the active state was not checked and signaled. So I added such a signal callback for 'activate' signal to update the  audio->monitoring_duplicating_source.
- Tests done: scene A with AOC 1 (on default device), switched to scene B with AOC 2 (on same device but set explicitly). Deduplication of a sine source is triggered and produces the expected level.
